core/src/test/scala/unit/kafka/log/LogTest.scala

```diff
@@ -1072,6 +1083,110 @@ class LogTest extends JUnitSuite {
 
     log.delete()
     assertEquals("The number of segments should be 0", 0, log.numberOfSegments)
-    assertEquals("The number of deleted segments shoud be zero.", 0, log.deleteOldSegments(_ => true))
+    assertEquals("The number of deleted segments should be zero.", 0, log.deleteOldSegments())
+  }
+
+
+  @Test
+  def shouldDeleteSizeBasedSegments() {
+    val set = TestUtils.singleMessageSet("test".getBytes)
+    val (_, log: Log) = createLog(set.sizeInBytes, retentionBytes = set.sizeInBytes * 10)
+
+    // append some messages to create some segments
+    for (i <- 0 until 15)
+      log.append(set)
+
+    log.deleteOldSegments
+    assertEquals("should have 2 segments", 2,log.numberOfSegments)
+  }
+
+  @Test
+  def shouldNotDeleteSizeBasedSegmentsWhenUnderRetentionSize() {
+    val set = TestUtils.singleMessageSet("test".getBytes)
+    val (_, log: Log) = createLog(set.sizeInBytes, retentionBytes = set.sizeInBytes * 15)
+
+    // append some messages to create some segments
+    for (i <- 0 until 15)
+      log.append(set)
+
+    log.deleteOldSegments
+    assertEquals("should have 3 segments", 3,log.numberOfSegments)
+  }
+
+  @Test
+  def shouldDeleteTimeBasedSegmentsReadyToBeDeleted() {
+    val set = TestUtils.singleMessageSet("test".getBytes, timestamp = 10)
+    val (myTime, log) = createLog(set.sizeInBytes, retentionMs = 10000)
+
+    // append some messages to create some segments
+    for (i <- 0 until 15)
+      log.append(set)
+
+    log.deleteOldSegments()
+    assertEquals("There should be 1 segment remaining", 1, log.numberOfSegments)
+  }
+
+  @Test
+  def shouldNotDeleteTimeBasedSegmentsWhenNoneReadyToBeDeleted() {
+    val set = TestUtils.singleMessageSet("test".getBytes, timestamp = System.currentTimeMillis())
```

Comment:
To be more reliable, perhaps we can move this line to after createLog and set the timestamp to be myMockTime.milliseconds?